openstack/datadog_checks/openstack/openstack.py

```diff
@@ -553,6 +558,13 @@ def __init__(self, name, init_config, agentConfig, instances=None):
         self.backoff = {}
         random.seed()
 
+        # Timestamp used to filter the call to get the list of nova servers
+        self.changes_since_time = {}
+
+        # Cache the list of server id's available
+        # This will have a key of instance_id, and a list of servers
+        self.servers = {}
```

Comment:
I think it's probably fine to 'split' the servers by instance id (although I'm not sure it's necessary - read ahead), but I think it would probably also make sure to index the servers by sever `id` or by `name`. This would allow you to get to a server's details in O(1), to verify state, etc.

ie. `self.servers[i_key][id]` or `self.servers[i_key][name]`. 

I understand that you need to keep track of the servers actually in use by each instance (ie. tenant/project), I'm wondering if it wouldn't be better for you to keep a separate cache, indexed by instance key, to a set of server ids available in a certain instance, and a separate cache for the actual server details. 

We should also ask ourselves if it makes sense to actually cache the servers by instance key (if a server can indeed belong to two different tenants or whatever, not sure it's possible). But even if it is possible, there's just the one instance which cannot be powered on and off at the same time. If we cache by instance then we'll have inconsistent states for the same VM in the cache. It's probably not the end of the world as we would get the right state for the VM in the next iteration. 

I just wanted to bring this up because it feels like it might be more efficient ways to index and iterate these caches which might get pretty bulky.